### ISSUE TITLE
Renaming same file a second time truncates and indexes filename

### ISSUE THREAD

**Issue by Unode** [jlegewie/zotfile#184]
Zotfile 4.1.5
Zotero 4.0.26.4
Using "Add suffix when authors are omitted" : `_et.al`
Using custom location for files and subfolder pattern (zotfile)
Using linked attachment base directory (zotero)

If I rename the same attachment twice in a row the first time I get the expected filename:
`Author_et.al-2015-This_title.pdf`
but a second rename causes:
`Author_et2.pdf`

This seems to be another form of #171 , however in this case the file is being treated as "duplicate" of itself.
I noticed this problem after a batch rename of an entire collection where some of its attachments had been previously renamed.

**Comment by jlegewie** [comment]
I will take a look at your issues but it might take a while...

**Comment by Unode** [comment]
If #207 gets accepted #117 should be fixed as well as the truncation aspect of this issue.

However I'm planning to submit another patch soon where a check is made and if the original and final file are one and the same, no rename action is performed. This should avoid the "toggle effect" that is seen after #207 when renaming the same file more than once.

**Comment by jlegewie** [comment]
Awesome, thanks! I will submit a new version to mozilla after you submit the other patch.

**Comment by Unode** [comment]
In any case, I'm having some issues with the latest master.
Getting errors when renaming files. Had to revert to the last release commit (v4.1.7) to be able to work on the patch.
The lastAuthor changes seem to be the cause of the issue. I can file a new issue if you prefer.

**Comment by jlegewie** [comment]
That would be great. It's possible that I didn't test them thoroughly when I accepted the pull request. The lack of good automatic tests... :(

**Comment by Unode** [comment]
Filed under #208

**Comment by Unode** [comment]
Hi again Joscha,

So the "toggle effect" I mentioned earlier is actually one with a silly cause. The code actually already checks if the original file and renamed file are one and the same. The problem is that in the "destination" path I ended up with something like: `/path/to/zotero/storage//zotfile/subfolder/file.pdf`. Notice the double `//`. This is compared to the actual location of the file which obviously doesn't include the double slash.

I went digging for the origin of the extra `/` and it turned out that it comes from what I used in the options under "Use subfolder defined by:". I had `/%w/%y`.
So `'/path/zotero' + this.folderSep + '/%w/%y'` caused the double `//`. And I think `this.folderSep` makes sense there.

The patch then would be about protecting users from this. I thought about patching the `updatePreferenceWindow` function to remove leading `/` or `\` from the subfolder if present. But this would cause problems with storage location since here the path may be absolute. I think this will also happen if storage includes a trailing `/`.

Patching the `completePath` function to remove duplicated `//` or `\\` could work but it's too hackish. It could fail badly in some unforeseen scenario like some character escape code.

So I'm a little unsure how to protect this. It seems any leading or trailing slash in the settings might cause problems. The exception being the custom storage location where a leading slash makes sense.

What are your thoughts on this?

**Comment by jlegewie** [comment]
`this.folderSep` is pretty old and it would be much better to handle this with  think the it would be better to use functions that join file path across platforms. I don't think there is anything in pure javascript but [OS.Path.join](https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/OSFile.jsm/OS.Path#OS.Path.join) should work. It still produces duplicate separators such as `//` but [`OS.Path.normalize `](https://developer.mozilla.org/en-US/docs/Mozilla/JavaScript_code_modules/OSFile.jsm/OS.Path#OS.Path.normalize) fixes that. I think combine these two would be the best solution or just using `OS.Path.normalize` if `OS.Path.join` implies to many changes. What do you think?



```
Components.utils.import("resource://gre/modules/osfile.jsm");
var path = OS.Path.join("/path/to/zotero/storage/", "zotfile/subfolder/file.pdf");
// still with double `//`
OS.Path.normalize(path)
// now the path looks good
```

**Comment by Unode** [comment]
`OS.Path.normalize(path)` looks good.

However `OS.Path.join` will not work as expected if the second argument starts with a `/`.

    > OS.Path.join("/path/to/zotero/storage/", "/zotfile/subfolder/file.pdf");
    "/zotfile/subfolder/file.pdf"

Perhaps just `normalize` will suffice and the `this.folderSep` approach is kept for now.

**Comment by jlegewie** [comment]
Yes, I agree. Consistently implementing `OS.Path.join` involves to many other changes.
